lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -816,3 +829,243 @@ def get_provider_vault_token(
+
+        certificate_from_provider, private_key = self._get_pki_intermediate_ca_from_relation()
+        if not certificate_from_provider or not private_key:
+            return
+        vault_service_ca_certificate = self._get_vault_service_ca_certificate()
+        if (
+            vault_service_ca_certificate
+            and vault_service_ca_certificate == certificate_from_provider.certificate
+        ):
+            if not self._intermediate_ca_exceeds_role_ttl(vault_service_ca_certificate):
+                self._tls_certificates_pki.renew_certificate(
+                    certificate_from_provider,
+                )
+                logger.debug("Renewing CA certificate")
+                return
+            logger.debug("CA certificate already set in the PKI secrets engine")
+            return
+        self._vault_pki.revoke_all_certificates()
+        self._vault_client.import_ca_certificate_and_key(
+            certificate=str(certificate_from_provider.certificate),
+            private_key=str(private_key),
+            mount=self._mount_point,
+        )
+        issued_certificates_validity = PKIManager.calculate_pki_certificates_ttl(
+            certificate_from_provider.certificate
+        )
+        if not self._vault_client.is_common_name_allowed_in_pki_role(
+            role=self._role_name,
+            mount=self._mount_point,
+            common_name=self._common_name,
+        ) or issued_certificates_validity != self._vault_client.get_role_max_ttl(
+            role=self._role_name, mount=self._mount_point
+        ):
+            self._vault_client.create_or_update_pki_charm_role(
+                allowed_domains=self._common_name,
+                mount=self._mount_point,
+                role=self._role_name,
+                max_ttl=f"{issued_certificates_validity}s",
+            )
+        self.make_latest_pki_issuer_default()
+
+    def make_latest_pki_issuer_default(self):
+        """Make the latest PKI issuer the default issuer.
+
+        This ensures that the latest issuer we have created is used for signing
+        certificates.
+        """
+        try:
+            first_issuer = self._vault_client.list_pki_issuers(mount=self._mount_point)[0]
+        except (VaultClientError, IndexError) as e:
+            logger.error("Failed to get the first issuer: %s", e)
+            return
+
+        try:
+            issuers_config = self._vault_client.read(path=f"{self._mount_point}/config/issuers")
+            if issuers_config and not issuers_config["default_follows_latest_issuer"]:
+                logger.debug("Updating issuers config")
+                self._vault_client.write(
+                    path=f"{self._mount_point}/config/issuers",
+                    data={
+                        "default_follows_latest_issuer": True,
+                        "default": first_issuer,
+                    },
+                )
+        except (TypeError, KeyError):
```

Comment:
Are we expecting those errors in the `read` or in the `write`. I would encourage making the try/except as small as possible so that the log messages in the except matches with the specific try - This helps in debugging.

Reply:
this code was moved from elsewhere. I can try to narrow it down if we feel that's worthwhile, but I was trying to avoid functional changes as much as possible.